Code/GraphMol/GeneralizedSubstruct/XQMol.cpp

```diff
@@ -0,0 +1,112 @@
+//
+//  Copyright (c) 2023, Greg Landrum and other RDKit contributors
+//
+//   @@ All Rights Reserved @@
+//  This file is part of the RDKit.
+//  The contents are covered by the terms of the BSD license
+//  which is included in the file license.txt, found at the root
+//  of the RDKit source tree.
+//
+//
+
+#include <cstdint>
+#include <variant>
+#include <sstream>
+#include <GraphMol/RDKitBase.h>
+#include <GraphMol/MolPickler.h>
+#include <GraphMol/MolBundle.h>
+#include <GraphMol/MolEnumerator/MolEnumerator.h>
+
+#include "XQMol.h"
+
+namespace RDKit {
+namespace GeneralizedSubstruct {
+
+ExtendedQueryMol::ExtendedQueryMol(const std::string &text, bool isJSON) {
+  if (!isJSON) {
+    initFromBinary(text);
+  } else {
+    initFromJSON(text);
+  }
+}
+
+std::vector<MatchVectType> SubstructMatch(
+    const ROMol &mol, const ExtendedQueryMol &query,
+    const SubstructMatchParameters &params) {
+  std::vector<MatchVectType> res;
+  if (std::holds_alternative<ExtendedQueryMol::RWMol_T>(query.xqmol)) {
+    res = RDKit::SubstructMatch(
+        mol, *std::get<ExtendedQueryMol::RWMol_T>(query.xqmol), params);
+#ifdef RDK_USE_BOOST_SERIALIZATION
+  } else if (std::holds_alternative<ExtendedQueryMol::MolBundle_T>(
+                 query.xqmol)) {
+    res = RDKit::SubstructMatch(
+        mol, *std::get<ExtendedQueryMol::MolBundle_T>(query.xqmol), params);
+  } else if (std::holds_alternative<ExtendedQueryMol::TautomerQuery_T>(
+                 query.xqmol)) {
+    res = std::get<ExtendedQueryMol::TautomerQuery_T>(query.xqmol)
+              ->substructOf(mol, params);
+  } else if (std::holds_alternative<ExtendedQueryMol::TautomerBundle_T>(
+                 query.xqmol)) {
+    const auto &vect =
+        std::get<ExtendedQueryMol::TautomerBundle_T>(query.xqmol);
+    for (const auto &tq : *vect) {
+      res = tq->substructOf(mol, params);
+      if (!res.empty()) {
+        break;
+      }
+    }
+#endif
+  } else {
+    UNDER_CONSTRUCTION("unrecognized type in ExtendedQueryMol");
+  }
+  return res;
+}
+
+ExtendedQueryMol createExtendedQueryMol(const RWMol &mol) {
+  auto bndl = MolEnumerator::enumerate(mol);
+  if (bndl.empty()) {
+    // nothing enumerated
+    auto tq = std::unique_ptr<TautomerQuery>(TautomerQuery::fromMol(mol));
+    if (tq->getTautomers().size() == 1) {
+      // no tautomers, return just the mol
+      return {std::make_unique<RWMol>(mol)};
```

Comment:
Do we want to call `adjustQueryProperties` on all of these mols before preforming substructure search, or am I missing something? If so, maybe it would be helpful to have `AdjustQueryParameters` as an optional parameter in this function

Reply:
oh, that's a really good suggestion. I'll add it.